src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs

```diff
@@ -113,7 +113,9 @@ protected virtual async Task OnConnectionComplete(IServiceConnection serviceConn
 
         private async Task RestartServiceConnectionCoreAsync(int index)
         {
-            await Task.Delay(GetRetryDelay(_defaultConnectionRetry));
+            // multiple concurrent connection attempts can quickly reach long back off delays
+            // so we adjust the effective retry count according to the number of connections
+            await Task.Delay(GetRetryDelay(_defaultConnectionRetry/FixedConnectionCount));
```

Comment:
I think after a recent change connection count does not always equal to FixedConnectionCount

Reply:
Also format the code

Reply:
@chenkennt On demand connections won't reconnect. This fix is OK for normal case.
@KKhurin I think it is better to connect server connections one by one, defend for a large `FixedConnectionCount` value (e.g. 6000, as the result we will try to create about 6000 connections per minute, 100 per second, it is still unreasonable reconnect rate).

Reply:
Make sense. Then what really needs to be fixed is to make this method synchronized?

Reply:
We shall use a queue to implement a producer-consumer model. When a connection failed, it just put itself(index) to the queue and use a seperate task to do the real reconnection and let it control the retry delay.

Reply:
I see. There are several requirements:
1. Reconnect all dropped fixed connections relatively fast in case of an intermittent short term outage. Currently a 5 seconds outage will result in most of the connections not recovering for a minute. We'd like all the connections to reconnect much faster, ideally with some predictable pattern.

2. Throttle the number of reconnect attempts in case of a large number of configured connections. Like @vwxyzh noted 100 retries for second is unacceptable (and unnecessary). I ran a quick test with the app server configured to have 1500 connections and retrying to connect to the service after a few seconds of outage consumed ~30 seconds of the process CPU time (~6000 retries).

How about this optimization: 
Only one retry happens each 1, 2, 4, 8, 16, 32, 60 seconds but as soon as a connection retry succeeds all other connections will stop waiting, all will try to connect, those which failed will start trying with the back off interval reset to 1s. In other words the exponentially increasing probing is performed with only 1 attempt per back off interval, but the back off interval gets reset as soon as a successful connection is made.